### ISSUE TITLE
Grad-CAM implementation with two data inputs

### ISSUE THREAD

**Issue by fraseralex96** [keras-team/keras-io#1452]
Hi Team,

I was looking for support online for this but couldn't find anyone that had produced an online tutorial for running grad-CAM with two data inputs.

My CNN architecture is as follows (see below), I have two inputs xo_train and xd_train, one is omic data and the other is one-hot encoded data. Would it be possible to run grad-CAM with this?

` # CNN architecure

    # set layer weights initialiser
    initializer = keras.initializers.GlorotUniform(seed=seed)
    
    # phosphoproteomics data input
    x_input = layers.Input(shape=(xo_train.shape[1],1))
    # 1st convolution layer
    x = layers.Conv1D(filters=8, kernel_size=4, kernel_initializer=initializer, activation='relu')(x_input) 
    x = layers.BatchNormalization()(x)
    x = layers.MaxPooling1D()(x)
    x = layers.Dropout(0.1)(x)
    # 2nd convolution layer
    x = layers.Conv1D(filters=16, kernel_size=4, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.MaxPooling1D()(x)
    x = layers.Dropout(0.1)(x)
    x = layers.Flatten()(x)

    # dense layers for xo_train
    x = layers.Dense(256, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.5)(x)
    x = layers.Dense(256, kernel_initializer=initializer, activation='relu')(x)
    x = layers.BatchNormalization()(x)
    x = layers.Dropout(0.5)(x)
    
    # one-hot encoded drug data input
    y_input = layers.Input(shape=(xd_train.shape[1]))
    #y = Dense(128, kernel_initializer=initializer, activation="relu")(y_input) 
    y = y_input
    
    # Concatenate phosphoproteomics and one-hot drug data + dense layers & activation layer
    z = layers.concatenate([x, y])
    z = layers.Dense(64, kernel_initializer=initializer, activation='relu')(z)
    z = layers.BatchNormalization()(z)
    z = layers.Dense(64, kernel_initializer=initializer, activation='relu')(z)
    z = layers.BatchNormalization()(z)
    z = layers.Dense(1, kernel_initializer=initializer)(z) # actiation layer is 1 neuron for single regression prediction value

    model = keras.Model([x_input, y_input], z)
    
    model.compile(optimizer=keras.optimizers.RMSprop(learning_rate=learning_rate, 
                                                        momentum=momentum), 
                                                        loss='mse', metrics=['mae'])`

**Comment by grasskin** [comment]
Hi @fraseralex96, Keras should be able to run Grad-CAM for multiple inputs. Follow https://keras.io/examples/vision/grad_cam/ but instead of using an existing model use yours. Note that your last convolutional layer does not directly take your `y_input`. I suggest visualizing Grad-CAM for only your `x_input` and convolutional layer instead of your full input. Your code is non-functional on our end due to missing data but it is my suspicion that the gradient will not be useful when taking into account your `y_input`. Leaving open for community follow-ups.

**Comment by fraseralex96** [comment]
Hi @grasskin,

Thank you for your response! I will let you know if I have any further questions.

Alex

**Comment by fraseralex96** [comment]
Hi,

I managed to implement grad-CAM with the imaging examples, however, I am wondering if it will be at all possible to repeat it for proteomic (tabular) data? My impression is that I will receive weightings for the importance of the feature maps of the last convolutional layer but since the omic data in a certain position in the input array is unrelated to data around it, this will be hard to then translate into importance of individual feature inputs. Am I correct or is there a way to use it for omic data?

Alex

**Comment by grasskin** [comment]
Hi @fraseralex96, I understand your intuition here for using GradCAM as a way to explain relative importance of features. However, GradCAM as a technique is designed to explain the activations of convolutional networks. I'm not quite sure what you're trying to do or with tabular data. I recommend you look into other techniques based on your use case. Best in class for tabular data is usually XGBoost. For interpretable models in general I would look into SHAP and  Shapley values.
